### PR TITLE
core: do not reset tty if there's no tty configured

### PR DESCRIPTION
**Pull request by bluca**
Otherwise a lock is taken, which can be taken by multiple units at the same time, depending on each other, causing a deadlock.

Fixes https://github.com/systemd/systemd/issues/29682

### CODE REVIEW

**Comment by bluca** [comment]
@mrc0mmand does this help with the efi.mount race?

File: src/core/execute.c [hunk 1]
```diff
@@ -116,7 +130,8 @@ void exec_context_tty_reset(const ExecContext *context, const ExecParameters *p)
 
         if (p && p->stdin_fd >= 0)
                 fd = p->stdin_fd;
```
Comment (poettering) [inline src/core/execute.c:132]: if we have an stdin_fd passed in, then we should also have a similar logic, i.e. if isatty() fails on the fd we should exit cleanly without doing anything.

File: src/core/execute.c [hunk 2]
```diff
@@ -116,7 +130,8 @@ void exec_context_tty_reset(const ExecContext *context, const ExecParameters *p)
...
-        else if (path) {
+        else if (context->tty_path || is_terminal_input(context->std_input) ||
+                        is_terminal_output(context->std_output) || is_terminal_output(context->std_error)) {
```
Comment (poettering) [inline src/core/execute.c]: would be nice to have a helper:

```c
static bool exec_context_uses_tty(const ExecContext *c) {
        …
````

that encapsulates this check.

but doesn't matter.

**Comment by poettering** [comment]
lgtm, just two minor additions

File: src/core/execute.c (hunk 1 above)
Comment (bluca) [reply src/core/execute.c:132]: added

**Comment by mrc0mmand** [comment]
> @mrc0mmand does this help with the efi.mount race?

I have it running for 30+ minutes in a loop and so far it hasn't failed once, so I'd say this indeed helps.

**Comment by bluca** [comment]
Great, thank you

**Comment by mrc0mmand** [comment]
Sanitizers seem to be unhappy:

```
../build/src/basic/terminal-util.c:333:27: runtime error: null pointer passed as argument 1, which is declared to never be null
/usr/include/fcntl.h:213:6: note: nonnull attribute specified here
```

**Comment by bluca** [comment]
Ah yes, should be fixed

**Comment by poettering** [comment]
lgtm! thanks!

**Pull request merged**